ops/snapshot_mainchain.go

```diff
@@ -0,0 +1,202 @@
+package ops
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"os"
+
+	"github.com/celer-network/goutils/log"
+	"github.com/celer-network/sgn/common"
+	"github.com/celer-network/sgn/mainchain"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+type Delegator struct {
+	DelegatedStake    *big.Int `json:"delegated_stake"`
+	UndelegatingStake *big.Int `json:"undelegating_stake"`
+}
+
+type Candidate struct {
+	StakingPool     *big.Int              `json:"staking_pool"`
+	CommissionRate  *big.Int              `json:"commission_rate"`
+	MinSelfStake    *big.Int              `json:"min_self_stake"`
+	RateLockEndTime *big.Int              `json:"rate_lock_end_time"`
+	Status          *big.Int              `json:"status"`
+	UnbondTime      *big.Int              `json:"unbond_time"`
+	Delegators      map[string]*Delegator `json:"delegators"`
+}
+
+type Snapshot struct {
+	EndBlockNumber uint64                `json:"end_block_number"`
+	Candidates     map[string]*Candidate `json:"candidates"`
+}
+
+const (
+	startBlkNum   = uint64(8580124)
+	queryInterval = uint64(200)
+
+	prevSnapshotFlag = "prev-snapshot"
+)
+
+var (
+	candidateMap = make(map[string]*Candidate)
+)
+
+func snapshotMainchain() error {
+	snapshot := Snapshot{
+		EndBlockNumber: startBlkNum,
+		Candidates:     candidateMap,
+	}
+	prevSnapshotFile := viper.GetString(prevSnapshotFlag)
+	if prevSnapshotFile != "" {
+		prevSnapshotBytes, err := ioutil.ReadFile(prevSnapshotFile)
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(prevSnapshotBytes, &snapshot)
+		if err != nil {
+			return err
+		}
+	}
+
+	ethClient, err := common.NewEthClientFromConfig()
+	if err != nil {
+		return err
+	}
+
+	header, err := ethClient.Client.HeaderByNumber(context.Background(), nil)
+	if err != nil {
+		return err
+	}
+
+	latestBlock := header.Number.Uint64()
+	dposContract := ethClient.DPoS
+	err = snapshotInitializeCandidate(dposContract, snapshot.EndBlockNumber, latestBlock)
+	if err != nil {
+		return err
+	}
+
+	for cur := snapshot.EndBlockNumber; cur < latestBlock; {
+		nextCur := cur + queryInterval
+		err = snapshotDelegate(dposContract, cur, nextCur)
```

Comment:
To be more efficient, we can query the `UpdateDelegatedStake` event instead to directly update both the delegator set and the stakes.